assign_rights/tests.py

```diff
@@ -95,21 +109,60 @@ def test_grouping_form(self):
 class TestRightsAssembler(TestCase):
     def setUp(self):
         add_rights_shells()
+        add_rights_acts()
         self.assembler = RightsAssembler()
+        self.rights_ids = [obj.pk for obj in RightsShell.objects.all()]
 
     def test_retrieve_rights(self):
         """Tests the retrieve_rights method.
 
         Asserts the method returns a list or DoesNotExist exception.
         """
-        rights_ids = [obj.pk for obj in RightsShell.objects.all()]
-        assembled = self.assembler.retrieve_rights(rights_ids)
+        assembled = self.assembler.retrieve_rights(self.rights_ids)
         self.assertTrue(isinstance(assembled, list))
-        self.assertEqual(len(rights_ids), len(assembled))
+        self.assertEqual(len(self.rights_ids), len(assembled))
 
-        rights_ids.append(len(rights_ids) + 1)
+        deleted = random.choice(RightsShell.objects.all())
+        deleted.delete()
+        self.rights_ids.append(deleted.pk)
         with self.assertRaises(RightsShell.DoesNotExist):
-            assembled = self.assembler.retrieve_rights(rights_ids)
+            assembled = self.assembler.retrieve_rights(self.rights_ids)
+
+    def test_calculate_dates(self):
+        """Tests the calculate_dates method.
+
+        Asserts the method returns a DateTimeField or None if the end date is open.
+        Asserts that the relative delta of the calculated date and the correct end date
+        is equal to the end date period of the object.
+        """
+        assembled = self.assembler.retrieve_rights(self.rights_ids)
+        # grouping_end_date = random_date() wasn't working here because it returned a string
+        now = date.today()
+        grouping_end_date = now - relativedelta(years=random.randint(2, 50))
+        grouping_end_date.isoformat()
+        for object in assembled:
+            object_date = self.assembler.calculate_dates(object, grouping_end_date)
+            if object.end_date_open:
+                self.assertEqual(object_date, None)
+            elif object.end_date:
+                self.assertEqual(relativedelta(object_date, object.end_date).years, object.end_date_period)
+                self.assertTrue(isinstance(object_date, date))
+            else:
+                self.assertEqual(relativedelta(object_date, grouping_end_date).years, object.end_date_period)
+                self.assertTrue(isinstance(object_date, date))
```

Comment:
These lines and 156-165 below look to be identical, so I'd suggest factoring them out to a separate function that gets called twice.